generate/avro.go

```diff
@@ -0,0 +1,344 @@
+	}
+	outPath := filepath.Join(dir, strings.ToLower(name)+".avsc")
+	specFile, err := os.Create(outPath)
+	if err != nil {
+		return outPath, fmt.Errorf("failed to open file %q: %v", outPath, err)
+	}
+
+	var writer io.Writer
+	var buf *bytes.Buffer
+	writer = specFile
+	if pretty {
+		buf = &bytes.Buffer{}
+		writer = buf
+	}
+
+	cfg := avroConfig{
+		dir:    dir,
+		writer: writer,
+	}
+
+	fmt.Fprint(cfg.writer, "{")
+	astutil.Apply(spec, writeAvroStruct(cfg, name, metadataFields), nil)
+	fmt.Fprint(cfg.writer, "]}")
+
+	if pretty {
+		rawJSON := json.RawMessage(buf.Bytes())
+		enc := json.NewEncoder(specFile)
+		enc.SetIndent("", "  ")
+		if err := enc.Encode(rawJSON); err != nil {
+			return "", fmt.Errorf("failed writing pretty output to file: %v", err)
+		}
+	}
+
+	if err := specFile.Close(); err != nil {
+		return "", fmt.Errorf("failed closing file: %v", err)
+	}
+
+	return outPath, nil
+}
+
+// parseGoStruct parses the go file at path returning the named struct type definition as an *ast.TypeSpec
+func parseGoStruct(name, path string) (*ast.TypeSpec, error) {
+	fileSet := token.NewFileSet()
+	goFile, err := parser.ParseFile(fileSet, path, nil, parser.AllErrors)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse generate go file %q: %v", path, err)
+	}
+	if !ast.FilterFile(goFile, func(itemName string) bool { return itemName == name }) {
+		return nil, fmt.Errorf("a struct named %q was not found in file %q", name, path)
+	}
+
+	if length := len(goFile.Decls); length != 1 {
+		return nil, fmt.Errorf("failed to filter declarations to a single one named %q, found %d", name, length)
+	}
+
+	d, ok := goFile.Decls[0].(*ast.GenDecl)
+	if !ok {
+		return nil, errors.New("filtered declaration is of unknown ast type")
+	}
+	if len(d.Specs) != 1 {
+		return nil, errors.New("unexpected number of specs in declaration")
+	}
+	t, ok := d.Specs[0].(*ast.TypeSpec)
+	if !ok {
+		return nil, errors.New("filtered type is of unknown ast type")
+	}
+
+	return t, nil
+}
+
+func parseStructTag(literal *ast.BasicLit) (name, description string, omitEmpty bool) {
+	if literal == nil {
+		return
+	}
+	tag := reflect.StructTag(strings.Trim(literal.Value, "`"))
+
+	description = tag.Get("description")
+	jsonValue := tag.Get("json")
+	jsonSplits := strings.Split(jsonValue, ",")
+	name = jsonSplits[0]
+	if len(jsonSplits) > 1 {
+		for _, split := range jsonSplits[1:] {
+			if strings.ToLower(split) == "omitempty" {
+				omitEmpty = true
+				return
+			}
+		}
+	}
+	return
+}
+
+// writeAvroStruct returns an apply function intended to be called for the start of each node.
+// The corresponding Post function will be called after the children of the node have been traversed.
+func writeAvroStruct(cfg avroConfig, name, defaultFields string) astutil.ApplyFunc {
+	return func(c *astutil.Cursor) bool {
+		if c.Name() == "Node" {
+			var namespace string
+			if len(cfg.namespace) != 0 {
+				namespace = fmt.Sprintf(`"namespace":%q,`, strings.Join(cfg.namespace, "."))
```

Comment:
Perhaps rename the `namespace` variable for clarity so that it's obvious it includes json formatting